### PR TITLE
feat(manager/regex)!: allow arbitrary regex groups for templates

### LINKED ISSUE #11387: Validate regex manager combination strategy
### Describe the proposed change(s).

The `combination` strategy seems buggy to me, as we itereate two times over all matches.
This should only happen once.

https://github.com/renovatebot/renovate/blob/8e3f184c717115b574213e49fd6618a140c85ae0/lib/manager/regex/index.ts#L154-L167

### PR DESCRIPTION
**Pull request by secustor**
## Changes:
This PR allows the usage of arbitrary capture groups inside of regex manager templates and further adds refactor which this allows.

BREAKING_CHANGE
Only regex managers using the `combination` matchStringStrategy are affected of this change!
Currently capture groups which are empty but still match the regex are ignored.     
This is no longer the case!    
Subsequent matchGroups will now overwrite previous ones, even if the later one is empty.     

## Context:
Fixes #11387

Related: #12279

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by rarkins** [comment]
@secustor can you resolve the merge conflict?

**Comment by secustor** [comment]
@rarkins Done.

**Comment by rarkins** [comment]
Is it possible to separate out the refactor from the breaking change? It's currently pretty hard to review due to the file changes.

**Comment by secustor** [comment]
I have reverted the addition of an `utils.ts`, which should make the changes clearer.    

But I'm not sure what is missing for codecov. In my overview there is no reference to the offending lines

**Comment by viceice** [comment]
> I have reverted the addition of an `utils.ts`, which should make the changes clearer.

The idea was to open a new pr with just the refactoring, so we can merge that before this.

> 
> But I'm not sure what is missing for codecov. In my overview there is no reference to the offending lines

I think this is only a codecov issue, as there are no coverage changes. 🤷‍♂️

File: lib/manager/regex/index.ts [hunk 1]
```diff
@@ -25,7 +25,10 @@ const validMatchFields = [
   'depType',
 ];
 
-const mergeFields = ['registryUrls', ...validMatchFields];
+interface ExtractionTemplate {
+  groups: Record<string, string>;
+  replaceString: string;
+}
```
Comment (viceice) [inline lib/manager/regex/index.ts:31]: please moved to `./types.ts`

**Review by viceice — COMMENTED** [review]
Do we need to change our own regex managers or will they work as before?

Looks good so far

**Comment by secustor** [comment]
If you are referring to https://github.com/renovatebot/.github/blob/main/default.json, then no changes are necessary as this only affects regex managers with matchStringStrategy `combination`.

**Review by viceice — DISMISSED** [review]

**Pull request merged**